boards: arm: Add support for NUCLEO-L496ZG

gon1332 opened the pull request:
Hi everyone,

I started using Zephyr for the past 3 months and although I'm learning, I feel that I have many things to learn. So, with this is pull request I am making my first steps to giving back to the community and also learn from you.

I will constantly add stuff little by little, and according to your guidelines, I hope this PR will become part of the project.

gon1332 committed:
boards: arm: Add initial support for NUCLEO-L496ZG

This commit adds the simplest configuration for this Nucleo-144 board
stripped down to the parts that are currently tested.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

galak commented on boards/arm/nucleo_l496zg/board.h:
```diff
@@ -0,0 +1,12 @@
+/*
```
Please remove board.h

galak reviewed (changes requested):
Please remove board.h

galak commented:
Also, missing board docs

vaussard commented on boards/arm/nucleo_l496zg/nucleo_l496zg_defconfig:
```diff
@@ -0,0 +1,41 @@
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=80000000
+
+# enable uart driver
+CONFIG_SERIAL=n
```
Comment contradicts the defconfig. Remove? But it would be good to have at least the console to start with.

vaussard commented on boards/arm/nucleo_l496zg/nucleo_l496zg_defconfig:
```diff
@@ -0,0 +1,41 @@
+
+# console
+CONFIG_CONSOLE=n
+CONFIG_UART_CONSOLE=n
```
Comment contradicts the defconfig. Remove?

vaussard commented on boards/arm/nucleo_l496zg/nucleo_l496zg_defconfig:
```diff
@@ -0,0 +1,41 @@
+CONFIG_UART_CONSOLE=n
+
+#enable pwm
+CONFIG_PWM=n
```
Comment contradicts the defconfig. Remove? Or add support (see for example nucleo_l476rg).

erwango replied:
Indeed, please remove.
Having this line here could say that this is available but not enabled, which is not the case as no pin is configured

gon1332 replied:
My next move is to support pwm. So, with the next commit it will be resolved. :)

erwango replied:
Ok, so ideally, group your changes into coherent commits.

vaussard commented on boards/arm/nucleo_l496zg/pinmux.c:
```diff
@@ -0,0 +1,29 @@
+#include <pinmux/stm32/pinmux_stm32.h>
+
+/* pin assignments for NUCLEO-L496ZG board */
+static const struct pin_config pinconf[] = {
```
Maybe you could add at least the pinmux for the UART console. Otherwise most Zephyr examples will not be very useful.

gon1332 replied:
@vaussard Console port would be my next step.

vaussard reviewed (commented):
Hello, some minor comments. It would be good to have at least a console :)

gon1332 committed:
boards: nucleo_l496zg: Add UART for console

Used LPUART 1 for console subsys. Also, added USART2 and USART3.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

gon1332 committed:
boards: nucleo_l496zg: Add docs & flashing support

Added documentation and openocd support.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

gon1332 commented:
@galak Documentation and openocd support has been added.

dbkinder reviewed (changes requested):
So close on the doc addition (thanks).  One formatting issue to fix.

gon1332 commented:
@galak I also removed board.h

dbkinder reviewed (approved):
+1 for docs, thanks.

gon1332 committed:
boards: nucleo_l496zg: Add pwm mapping and RTC

Supported PWM at pins A0, D3, D5 and D6 which are also Arduino
compatible. Also the onboard red LED can be used with PWM.

The basic/{fade_led, blink_led} samples to apply to this board
seemlessly and operate on the onboard red LED.

Also supported RTC.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

gon1332 commented:
Yesterday I've spoiled a bit my history after the rebase, so I'm in the process of restoring it.

gon1332 commented:
@galak @erwango @dbkinder Please review my newest additions. 

The next step is to add SPI and I2C. What else do you think is needed?

gon1332 committed:
boards: nucleo_l496zg: Add SPI, I2C

Added Aduino SPI and I2C mappings.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

erwango commented on samples/basic/blink_led/src/main.c:
```diff
@@ -17,7 +17,8 @@
 	defined(CONFIG_SOC_STM32F413XX) || defined(CONFIG_SOC_STM32L476XG) || \
-	defined(CONFIG_SOC_STM32F407XG) || defined(CONFIG_SOC_STM32F302X8)
+	defined(CONFIG_SOC_STM32F407XG) || defined(CONFIG_SOC_STM32F302X8) || \
+	defined(CONFIG_SOC_STM32L496XG)
```
Remove this and add pwm-led0 alias to the board.
Cf hexiwear_k64 board

gon1332 replied:
Nice, wasn't aware of pwm-leds.

gon1332 replied:
@erwango For this to happen I think I have to add `pwmX` in [dts/arm/st/stm32l4.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/st/l4/stm32l4.dtsi) like in [dts/arm/nxp/nxp_k6x.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/arm/nxp/nxp_k6x.dtsi).

Or is there any way to write the `&timers2:pwm` with correct syntax and avoid the above work?

    ...
        pwmleds {
            compatible = "pwm-leds";

            pwm_led: pwm_led {
                pwms = <&timers2:pwm 0 40000000>;
            };
        };
    ...
    };

    &timers2 {
        status = "ok";

        pwm {
            status = "ok";
        }
    };

erwango replied:
@galak , since you introduced pwm-leds, any idea?

galak replied:
> @galak , since you introduced pwm-leds, any idea?

I'm not aware of any syntax support for doing this.

galak replied:
You could do: `pwms = <&{/soc/timers@40000000/pwm} 3 4>;`

erwango replied:
@gon1332 , are you good with this answer?

gon1332 replied:
It's something I was looking for. I will put it into code ASAP and will inform you.

gon1332 replied:
@erwango Thanks to @galak's the comment, now Nucleo L496ZG has an on-board red PWM LED.
I omitted the green and blue ones, since I should add the corresponding ST timers, and I do not have time for this at the moment.

Please review my work.

erwango commented on boards/arm/nucleo_l496zg/nucleo_l496zg_defconfig:
```diff
@@ -35,7 +35,7 @@ CONFIG_CONSOLE=y
 
 #enable pwm
-CONFIG_PWM=n
+CONFIG_PWM=y
```
PWM activation is application responsibility.
Remove this line

gon1332 commented:
Also rebased to master.

gon1332 commented:
@galak @erwango 
Hi there! Because the pull request is stalled at this point, but it is working, do you think of a possible merge. Since I and my colleagues work on Nucleos, further additions can be made on another PR in the future.

erwango commented:
@gon1332 ,

If you don't mind, I'd like to wait feedback from @galak on the PWM point.
Would be great to integrate the PWM device tree thing, which should be the new reference.
@galak, can you check the upper question from @gon1332 ?

gon1332 committed:
boards: nucleo_l496zg: Use shell-uart

Added shell-uart for this board.

Signed-off-by: Ioannis Konstantelias <ikonstadel@gmail.com>

erwango commented:
@galak , aside from the rebase needed, can you check @gon1332 question above?

gon1332 commented:
@erwango I've rebased.

What workflow should I follow in order to work another Zephyr feature that needs this PR as a testing device?

Should I create a branch on top of this one? Something else?

erwango commented:
@gon1332 

> What workflow should I follow in order to work another Zephyr feature that needs this PR as a testing device?

You can create a new PR that is based on this one. So indeed on your side, this would be a new branch on top of that one

gon1332 commented:
@erwango Thanks to @galak's the comment, now Nucleo L496ZG has an on-board red PWM LED.
I omitted the green and blue ones, since I should add the corresponding ST timers, and I do not have time for this at the moment.

Please review my work.

gon1332 commented:
Has anything changed regarding the UART port configuration?

The tests now fail due to:

> warning: UART_STM32_PORT_2 (defined at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/boards/arm/nucleo_l496zg/Kconfig.defconfig:15) defined without a type

which is something new.

gon1332 commented:
> Has anything changed regarding the UART port configuration?
> 
> The tests now fail due to:
> 
> > warning: UART_STM32_PORT_2 (defined at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/boards/arm/nucleo_l496zg/Kconfig.defconfig:15) defined without a type
> 
> which is something new.

OK, I should rebase, in order to solve it. I just saw this one https://github.com/zephyrproject-rtos/zephyr/pull/11234.

gon1332 commented:
> > Has anything changed regarding the UART port configuration?
> > The tests now fail due to:
> > > warning: UART_STM32_PORT_2 (defined at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/boards/arm/nucleo_l496zg/Kconfig.defconfig:15) defined without a type
> > 
> > 
> > which is something new.
> 
> OK, I should rebase, in order to solve it. I just saw this one #11234.

Everything is OK. :)

erwango reviewed (commented):
I'm good with the change.
Please merge your 2 pwm commits and we'll be good.

gon1332 commented:
@erwango I merged the two commits and rebased as you suggested.

As for the failing test, it originates from another recent, independent merge on master.

erwango commented:
recheck

erwango commented:
> As for the failing test, it originates from another recent, independent merge on master.

It should be solved by now. CI should be passed in few minutes.
I've just spotted a very last (I hope) point which I'm having hard time to see where it have been introduced. I guess github is not presenting commits in the right order.
About the CONFIG_UART_X recent naming change, please use the new naming upfront, rather than introducing the deprecated CONFIG_UART_STM32_PORT_X then fixing it with a latter commit.
Sorry I'm not able to spot the exact code location, but you get the idea.

gon1332 commented:
> It should be solved by now. CI should be passed in few minutes.
I've just spotted a very last (I hope) point which I'm having hard time to see where it have been introduced. I guess github is not presenting commits in the right order.
About the CONFIG_UART_X recent naming change, please use the new naming upfront, rather than introducing the deprecated CONFIG_UART_STM32_PORT_X then fixing it with a latter commit.
Sorry I'm not able to spot the exact code location, but you get the idea.

Locally and through an android app, I can see the commits in the right order. For some reason Github reorders them. OK, I've got it.

erwango commented on boards/arm/nucleo_l496zg/Kconfig.defconfig:
```diff
@@ -0,0 +1,63 @@
+config BOARD
+	default "nucleo_l496zg"
+
+config UART_3
```
This one should not be here, better under if SERIAL. But I think we should remove it completly

gon1332 replied:
I agree with removing `UART_3`.

erwango replied:
Seems you still need to remove it from here

gon1332 replied:
As I see the Kconfig.defconfig, it has been removed. `UART_3` is removed only from this file and not from pinmux.c. Is that you are talking about?

erwango reviewed (changes requested):
@gon1332 , it seems that similarly to https://github.com/zephyrproject-rtos/zephyr/issues/12092, default serial port is lpuart1 rather than uart2.
How is that going for this board?

gon1332 commented:
@erwango I will apply your enhancement(https://github.com/zephyrproject-rtos/zephyr/pull/12177) regarding VCOM once it's merged.

erwango commented:
> @erwango I will apply your enhancement(#12177) regarding VCOM once it's merged.

So don't hesitate to +1 if you happen to test it

gon1332 commented:
1) Updated the Copyright text
2) Rebased to upstream/master

gon1332 commented:
@erwango I finally addressed your #12177 in this PR.

erwango commented:
@gon1332 

> @erwango I finally addressed your #12177 in this PR.

This is great but please can you fix the initial commit rather than doing it as an add-on?
Let's try to get this merge before end of week.

gon1332 commented:
> @gon1332
> 
> > @erwango I finally addressed your #12177 in this PR.
> 
> This is great but please can you fix the initial commit rather than doing it as an add-on?
> Let's try to get this merge before end of week.

Done!

erwango commented on samples/basic/fade_led/README.rst:
```diff
@@ -28,7 +28,8 @@ Connect PWM1(PA8) to LED
 No special board setup is necessary because there is an on-board RGB LED
-connected to the K64 PWM.
+connected to the K64 PWM and three on-board leds (red, green, blue) connected
+to the Nucleo's PWM.
```
You should not mention nucleo in Hexiwear section

erwango commented on boards/arm/nucleo_l496zg/Kconfig.defconfig:
```diff
@@ -24,6 +24,23 @@ config LPUART_1
+	default y
+
+endif # SPI
+
```
RTC is not added, please update commit message

erwango commented on boards/arm/nucleo_l496zg/Kconfig.defconfig:
```diff
@@ -17,11 +17,11 @@ config LPUART_1
 
 endif # UART_CONSOLE
 
-if UART_SERIAL
```
Fix this in the uart related commit.

erwango reviewed (changes requested):
3 issues to fix and this will be good from my side. Some inconsistencies remain in the way the commits are built (lines added in one commit and fixed in later one, files modified in wrong commit, ...), but no functional impact and we should move on. Please be more methodological next time.

gon1332 commented:
@erwango Thanks for the review. Some things were off due to the multiple rebases. I've addressed your requests.

erwango reviewed (changes requested):
Don't know if I missed these ones earlier, my bad if this is the case

gon1332 commented:
@erwango Fixed. Also merged a redundant commit to improve the commit flow.

erwango commented on samples/basic/blink_led/README.rst:
```diff
@@ -25,10 +25,17 @@ Nucleo_F103RB
 No special board setup is necessary because there is an on-board RGB LED
-connected to the K64 PWM.
+connected to the K64 PWM and three on-board leds (red, green, blue) connected
+to the Nucleo's PWM.
```
Still nucleo comment under hexiwear section

gon1332 replied:
Fixed

erwango reviewed (changes requested):
last one

erwango commented on boards/arm/nucleo_l496zg/Kconfig.board:
```diff
@@ -1,6 +1,6 @@
 # Kconfig - STM32L496ZG Nucleo board configuration
 #
-# Copyright (c) 2018 Ioannis Konstantelias <ikonstadel@gmail.com>
```
Please fix in initial commit rather than adding a fixing commit

gon1332 replied:
Done.

erwango commented on boards/arm/nucleo_l496zg/Kconfig.defconfig:
```diff
@@ -1,6 +1,6 @@
 # Kconfig - STM32L496ZG Nucleo board configuration
 #
-# Copyright (c) 2018 Ioannis Konstantelias <ikonstadel@gmail.com>
+# Copyright (c) 2018 Centaur Analytics, Inc
```
ditto

erwango commented on boards/arm/nucleo_l496zg/nucleo_l496zg.dts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Ioannis Konstantelias <ikonstadel@gmail.com>
+ * Copyright (c) 2018 Centaur Analytics, Inc
```
ditto

gon1332 commented:
@erwango Should we try to merge it?

gon1332 commented:
@galak Any news on this PR?

gon1332 commented:
@galak Also rebased to master.
